src/lsm/lsm_cursor.c

```diff
@@ -902,6 +902,107 @@ err:	__clsm_leave(clsm);
 }
 
 /*
+ * __clsm_random_chunk --
+ *	Pick a chunk at random, weighted by the size of all chunks. Weighting
+ * proportional to documents avoids biasing towards small chunks. Then open a
+ * random cursor on the chunk we have picked.
+ */
+static int
+__clsm_random_chunk(WT_SESSION_IMPL *session,
+    WT_CURSOR_LSM *clsm, WT_CURSOR **cursor)
+{
+	WT_DECL_RET;
+	WT_LSM_CHUNK *random_chunk;
+	WT_LSM_TREE *lsm_tree;
+	uint64_t checked_docs, i, rand_doc, total_docs;
+	const char *cfg[3];
+
+	cfg[0] = WT_CONFIG_BASE(session, WT_SESSION_open_cursor);
+	cfg[1] = "next_random=true";
+	cfg[2] = NULL;
+
+	random_chunk = NULL;
+	lsm_tree = clsm->lsm_tree;
+	rand_doc = __wt_random(&session->rnd);
+	__wt_lsm_tree_readlock(session, lsm_tree);
+	for (total_docs = i = 0; i < clsm->nchunks; i++) {
+		total_docs += lsm_tree->chunk[i]->count;
+	}
+
+	rand_doc %= total_docs;
+
+	for (checked_docs = i = 0; i < clsm->nchunks; i++) {
+		checked_docs += lsm_tree->chunk[i]->count;
+		if (rand_doc <= checked_docs) {
+			random_chunk = lsm_tree->chunk[i];
+			break;
+		}
+	}
+	ret = __wt_open_cursor(session, random_chunk->uri, NULL, cfg, cursor);
+	__wt_lsm_tree_readunlock(session, lsm_tree);
+	return (ret);
+}
+
+/*
+ * __clsm_next_random --
+ *	WT_CURSOR->next method for the LSM cursor type when configured with
+ * next_random.
+ */
+static int
+__clsm_next_random(WT_CURSOR *cursor)
+{
+	WT_CURSOR_LSM *clsm;
+	WT_CURSOR *c;
+	WT_DECL_RET;
+	WT_SESSION_IMPL *session;
+
+	c = NULL;
+	clsm = (WT_CURSOR_LSM *)cursor;
+
+	CURSOR_API_CALL(cursor, session, next, NULL);
+	WT_CURSOR_NOVALUE(cursor);
+	WT_ERR(__clsm_enter(clsm, false, false));
+
+	do {
+		WT_WITH_SCHEMA_LOCK(session, ret,
+		    WT_ERR(__clsm_random_chunk(session, clsm, &c)));
+		ret = c->next(c);
+		/*
+		 * Sometimes we may run on an empty chunk or for some reason be
+		 * unable to get a random doc. That's okay and we retry.
+		 */
+		if (ret == WT_NOTFOUND) {
+			WT_ERR(c->close(c));
+			ret = WT_NOTFOUND;
+			continue;
+		}
+
+		if (ret == 0)
+			ret = c->get_key(c, &cursor->key);
+		WT_ERR(c->close(c));
+
+		/*
+		 * Search back through the tree from the top to resolve any
+		 * updates or tombstones. If we find a tombstone, then pick a
+		 * new random doc and try again. We retry forever as eventually
+		 * we should see a current doc or the tombstones will get
+		 * merged out.
+		 */
+		ret = __clsm_lookup(clsm, &cursor->value);
+		if (ret != WT_NOTFOUND)
+			WT_ERR(ret);
+	} while (ret == WT_NOTFOUND);
```

Comment:
Can you explain what happens if `next_random` is called on an empty LSM tree?  It looks like it will get stuck in an infinite loop...